Indexing/CodeTree.cpp

```diff
@@ -357,109 +360,137 @@ bool CodeTree::CodeOp::equalsForOpMatching(const CodeOp& o) const
   }
 }
 
-CodeTree::SearchStruct* CodeTree::CodeOp::getSearchStruct()
+CodeTree::SearchStruct* CodeTree::CodeOp::getSearchStruct() const
 {
-  //the following line gives warning for not being according
-  //to the standard, so we have to work around
-//  static const size_t opOfs=offsetof(SearchStruct,landingOp);
-  static const size_t opOfs=reinterpret_cast<size_t>(
-	&reinterpret_cast<SearchStruct*>(8)->landingOp)-8;
-
-  SearchStruct* res=reinterpret_cast<SearchStruct*>(
-      reinterpret_cast<size_t>(this)-opOfs);
+  ASS(isSearchStruct());
+  return GET_CONTAINING_OBJECT(CodeTree::SearchStruct,landingOp,this);
+}
 
-  return res;
+std::ostream& operator<<(std::ostream& out, const CodeTree::CodeOp& op)
+{
+  switch (op.instrPrefix()) {
+  case CodeTree::SUCCESS_OR_FAIL:
+    if (op.isSuccess()) {
+      out << "success";
+    } else {
+      out << "fail";
+    }
+    break;
+  case CodeTree::LIT_END:
+    out << "lit end";
+    break;
+  case CodeTree::CHECK_GROUND_TERM:
+    out << "check ground term " << *op.getTargetTerm();
+    break;
+  case CodeTree::SUFFIX_INSTR:
+    switch(op.instrSuffix()) {
+    case CodeTree::CHECK_FUN:
+      out << "check fun " << env.signature->getFunction(op.arg())->name();
+      break;
+    case CodeTree::ASSIGN_VAR:
+      out << "assign var X" << op.arg();
+      break;
+    case CodeTree::CHECK_VAR:
+      out << "check var X" << op.arg();
+      break;
+    case CodeTree::SEARCH_STRUCT:
+      out << "search struct ";
+      auto ss = op.getSearchStruct();
+      switch(ss->kind) {
+        case CodeTree::SearchStruct::FN_STRUCT: {
+          auto fn_ss = static_cast<CodeTree::FnSearchStruct*>(ss);
+          out << "length " << fn_ss->length();
+          for (unsigned i = 0; i < fn_ss->length(); i++) {
+            out << " " << fn_ss->values[i] << " ";
+            if (fn_ss->targets[i]) {
+              out << *fn_ss->targets[i];
+            } else {
+              out << "nullptr";
+            }
+          }
+          break;
+        }
+        case CodeTree::SearchStruct::GROUND_TERM_STRUCT: {
+          auto gt_ss = static_cast<CodeTree::GroundTermSearchStruct*>(ss);
+          out << "length " << gt_ss->length();
+          for (unsigned i = 0; i < gt_ss->length(); i++) {
+            out << " " << *gt_ss->values[i] << " ";
+            if (gt_ss->targets[i]) {
+              out << *gt_ss->targets[i];
+            } else {
+              out << "nullptr";
+            }
+          }
+          break;
+        }
+      }
+      break;
+    }
+    break;
+  }
+  return out;
 }
 
-CodeTree::SearchStruct::SearchStruct(Kind kind)
+CodeTree::SearchStruct::SearchStruct(Kind kind, size_t length)
 : kind(kind)
 {
   landingOp.setAlternative(0);
   landingOp.setLongInstr(SEARCH_STRUCT);
-}
+  ASS(length);
 
-void CodeTree::SearchStruct::destroy()
```

Comment:
Not needed any more?

Reply:
Good catch, looks like I slipped here. Probably should reinstate this altogether.

Reply:
I think introducing virtual destructors for this does not worth it, right?

Reply:
Depends - I think you just `delete` whatever instead, right? As long as you get the deletion logic right, I'm fine with it.

Virtual destructors: only if you need different destruction logic per-class. I think here it's just deleting the memory it occupied.